File: frontend/src/api/series.ts
/**
 * Read-only client for `/api/series/{id}` (sub-phase 11d).
 *
 * Mirrors the response DTO in `backend/src/models/series.rs`. Snake_case
 * field names match the wire shape; the Zod schema is the boundary
 * guard per `frontend/CLAUDE.md`.
 *
 * The frontend computes the "completeness" indicator (`own / total`)
 * from the response — `total = works.length`, `own = works.filter(w =>
 * w.manifestations.length > 0).length`. Works the user has no copy of
 * surface with an empty `manifestations` array.
 */
import { z } from "zod";

import { apiFetch } from "./fetch";

const WorkManifestationSchema = z.object({
  id: z.string(),
  isbn_13: z.string().nullable(),
  isbn_10: z.string().nullable(),
  cover_url: z.string(),
  ingestion_status: z.enum(["pending", "processing", "complete", "failed", "skipped"]),
  validation_status: z.string(),
  enrichment_status: z.enum(["pending", "in_progress", "complete", "failed", "skipped"]),
  created_at: z.string(),
});
/** One manifestation row attached to a work in a series response. */
export type SeriesWorkManifestation = z.infer<typeof WorkManifestationSchema>;

const SeriesWorkSchema = z.object({
  id: z.string(),
  title: z.string(),
  authors: z.array(z.string()),
  position: z.number().nullable(),
  manifestations: z.array(WorkManifestationSchema),
});
/** One work in a series. Empty `manifestations` array = the user owns no copy ("gap"). */
export type SeriesWork = z.infer<typeof SeriesWorkSchema>;

const SeriesDetailSchema = z.object({
  id: z.string(),
  name: z.string(),
  sort_name: z.string(),
  works: z.array(SeriesWorkSchema),
});
/** `GET /api/series/{id}` response envelope. */
export type SeriesDetail = z.infer<typeof SeriesDetailSchema>;

/**
 * Fetch a series detail by id. RLS-hidden rows resolve to 404
 * (existence-not-leaked).
 */
export async function getSeries(id: string, signal?: AbortSignal): Promise<SeriesDetail> {
  const body = await apiFetch(
    `/api/series/${encodeURIComponent(id)}`,
    signal ? { method: "GET", signal } : { method: "GET" },
  );
  return SeriesDetailSchema.parse(body);
}

File: backend/src/models/series.rs
//! Response DTOs for `/api/v1/series/{id}`.
//!
//! Wire-format conventions follow the JSON-API conventions ADR
//! (`adr/2026-05-22-json-api-conventions.md`): snake_case field names,
//! `Option<T>` for nullable, RFC 3339 timestamps.
//!
//! # RLS posture
//!
//! `series` and `series_works` carry no RLS policies (catalog data,
//! not user-scoped). RLS lives on `manifestations`, so per-work
//! `manifestations` slots come back empty for works the caller has no
//! visibility into. The route handler treats "every work in the
//! series has zero visible manifestations" as 404 to keep the
//! existence-not-leaked invariant intact for child accounts and
//! cross-user adult isolation.

use serde::Serialize;
use uuid::Uuid;

use crate::models::library::WorkManifestation;

/// `/api/v1/series/{id}` response. Carries the series identity plus an
/// ordered list of works; each work surfaces only the manifestations
/// the caller can see, so the frontend can render a completeness
/// indicator (`own = works with len(manifestations) > 0`, `total =
/// works.len()`).
#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
#[non_exhaustive]
pub struct SeriesDetail {
    /// `series.id`.
    pub id: Uuid,
    /// `series.name` (display).
    pub name: String,
    /// `series.sort_name` (collation key — surfaced so the UI can
    /// alpha-sort a list of series without a follow-up request).
    pub sort_name: String,
    /// Works in the series, ordered by `series_works.position` (NULLS
    /// LAST) then `works.id` for tiebreaking determinism.
    pub works: Vec<SeriesWork>,
}

/// One work within a series. Includes every manifestation of the work
/// the caller can see; an empty `manifestations` list means the user
/// does not yet own a copy ("gap" in the series).
#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
#[non_exhaustive]
pub struct SeriesWork {
    /// `works.id`.
    pub id: Uuid,
    /// `works.title`.
    pub title: String,
    /// Author display names ordered by `work_authors.position`.
    pub authors: Vec<String>,
    /// `series_works.position`; `None` when the membership row has a
    /// null position.
    pub position: Option<f64>,
    /// Manifestations of this work visible to the caller. Empty when
    /// the user owns no copy.
    pub manifestations: Vec<WorkManifestation>,
}
